**Import Libraries**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

**Import data**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
#Import data from gg drive
df = pd.read_excel('/content/drive/MyDrive/E-commerce/E Commerce churn.xlsx', sheet_name = 'E Comm')

# **Data cleaning**

In [ ]:
df_1 = df.copy()

In [ ]:
df_1.shape

(5630, 20)

In [ ]:
df_1.head(10)

CustomerID  Churn  Tenure PreferredLoginDevice  CityTier  WarehouseToHome  \
0       50001      1     4.0         Mobile Phone         3              6.0   
1       50002      1     NaN                Phone         1              8.0   
2       50003      1     NaN                Phone         1             30.0   
3       50004      1     0.0                Phone         3             15.0   
4       50005      1     0.0                Phone         1             12.0   
5       50006      1     0.0             Computer         1             22.0   
6       50007      1     NaN                Phone         3             11.0   
7       50008      1     NaN                Phone         1              6.0   
8       50009      1    13.0                Phone         3              9.0   
9       50010      1     NaN                Phone         1             31.0   

  PreferredPaymentMode  Gender  HourSpendOnApp  NumberOfDeviceRegistered  \
0           Debit Card  Female             3.0                         3   
1                  UPI    Male             3.0                         4   
2           Debit Card    Male             2.0                         4   
3           Debit Card    Male             2.0                         4   
4                   CC    Male             NaN                         3   
5           Debit Card  Female             3.0                         5   
6     Cash on Delivery    Male             2.0                         3   
7                   CC    Male             3.0                         3   
8             E wallet    Male             NaN                         4   
9           Debit Card    Male             2.0                         5   

     PreferedOrderCat  SatisfactionScore MaritalStatus  NumberOfAddress  \
0  Laptop & Accessory                  2        Single                9   
1              Mobile                  3        Single                7   
2              Mobile                  3        Single                6   
3  Laptop & Accessory                  5        Single                8   
4              Mobile                  5        Single                3   
5        Mobile Phone                  5        Single                2   
6  Laptop & Accessory                  2      Divorced                4   
7              Mobile                  2      Divorced                3   
8              Mobile                  3      Divorced                2   
9              Mobile                  3        Single                2   

   Complain  OrderAmountHikeFromlastYear  CouponUsed  OrderCount  \
0         1                         11.0         1.0         1.0   
1         1                         15.0         0.0         1.0   
2         1                         14.0         0.0         1.0   
3         0                         23.0         0.0         1.0   
4         0                         11.0         1.0         1.0   
5         1                         22.0         4.0         6.0   
6         0                         14.0         0.0         1.0   
7         1                         16.0         2.0         2.0   
8         1                         14.0         0.0         1.0   
9         0                         12.0         1.0         1.0   

   DaySinceLastOrder  CashbackAmount  
0                5.0          159.93  
1                0.0          120.90  
2                3.0          120.28  
3                3.0          134.07  
4                3.0          129.60  
5                7.0          139.19  
6                0.0          120.86  
7                0.0          122.93  
8                2.0          126.83  
9                1.0          122.93

**Data information**

Data has 20 columns and 5630 rows




*   CustomerID: Unique customer ID

*  Churn:      Churn Flag (1 is churned, 0 is not churned)

*   Tenure: Tenure of customer in organization



*   CityTier: City Tier




*   PreferredLoginDevice: Preferred login device of customer

*   WarehouseToHome: Distance in between warehdwouse to home of customer
*   PreferredPaymentMode: Preferred payment method of customer


*   Gender: Gender of customer


*   HourSpendOnApp: Number of hours spend on mobile application or website

*   NumberOfDeviceRegistered: Total number of devices is registered on particular customer


*   PreferedOrderCat: Preferred order category of customer in last month


*   SatisfactionScore: Satisfactory score of customer on service


*   MaritalStatus: Marital status of customer

*  NumberOfAddress: Total number of added added on particular customer
*   Complain: Any complaint has been raised in last month

*   OrderAmountHikeFromlastYear: Percentage increases in order from last year


*   CouponUsed: Total number of coupon has been used in last month



*   OrderCount: Total number of orders has been places in last month

*   DaySinceLastOrder: Day Since last order by customer

*   CashbackAmount: Average cashback in last month


















In [ ]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Data columns (total 19 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Churn                        5630 non-null   int64  
 1   Tenure                       5366 non-null   float64
 2   PreferredLoginDevice         5630 non-null   object 
 3   CityTier                     5630 non-null   int64  
 4   WarehouseToHome              5379 non-null   float64
 5   PreferredPaymentMode         5630 non-null   object 
 6   Gender                       5630 non-null   object 
 7   HourSpendOnApp               5375 non-null   float64
 8   NumberOfDeviceRegistered     5630 non-null   int64  
 9   PreferedOrderCat             5630 non-null   object 
 10  SatisfactionScore            5630 non-null   int64  
 11  MaritalStatus                5630 non-null   object 
 12  NumberOfAddress              5630 non-null   int64  
 13  Complain          

In [ ]:
#Null Values
df_1.isna().sum()

CustomerID                       0
Churn                            0
Tenure                         264
PreferredLoginDevice             0
CityTier                         0
WarehouseToHome                251
PreferredPaymentMode             0
Gender                           0
HourSpendOnApp                 255
NumberOfDeviceRegistered         0
PreferedOrderCat                 0
SatisfactionScore                0
MaritalStatus                    0
NumberOfAddress                  0
Complain                         0
OrderAmountHikeFromlastYear    265
CouponUsed                     256
OrderCount                     258
DaySinceLastOrder              307
CashbackAmount                   0
dtype: int64

**Check duplicate**

In [ ]:
#Check duplicate
duplicate_row = df_1[df_1.duplicated()]
len(duplicate_row)

556

Data had no duplicate values

**Handle Null Values**

In [ ]:
df_1.isna().sum()

Churn                            0
Tenure                         264
PreferredLoginDevice             0
CityTier                         0
WarehouseToHome                251
PreferredPaymentMode             0
Gender                           0
HourSpendOnApp                 255
NumberOfDeviceRegistered         0
PreferedOrderCat                 0
SatisfactionScore                0
MaritalStatus                    0
NumberOfAddress                  0
Complain                         0
OrderAmountHikeFromlastYear    265
CouponUsed                     256
OrderCount                     258
DaySinceLastOrder              307
CashbackAmount                   0
dtype: int64

In [ ]:
del df_1['CustomerID']

Before we handle missing values, let's see statistical summary of this dataset

In [ ]:
#Columns had missing values
column_had_miss = df_1.loc[:,df_1.isnull().any()]
column_had_miss.describe()

Tenure  WarehouseToHome  HourSpendOnApp  \
count  5366.000000      5379.000000     5375.000000   
mean     10.189899        15.639896        2.931535   
std       8.557241         8.531475        0.721926   
min       0.000000         5.000000        0.000000   
25%       2.000000         9.000000        2.000000   
50%       9.000000        14.000000        3.000000   
75%      16.000000        20.000000        3.000000   
max      61.000000       127.000000        5.000000   

       OrderAmountHikeFromlastYear   CouponUsed   OrderCount  \
count                  5365.000000  5374.000000  5372.000000   
mean                     15.707922     1.751023     3.008004   
std                       3.675485     1.894621     2.939680   
min                      11.000000     0.000000     1.000000   
25%                      13.000000     1.000000     1.000000   
50%                      15.000000     1.000000     2.000000   
75%                      18.000000     2.000000     3.000000   
max                      26.000000    16.000000    16.000000   

       DaySinceLastOrder  
count        5323.000000  
mean            4.543491  
std             3.654433  
min             0.000000  
25%             2.000000  
50%             3.000000  
75%             7.000000  
max            46.000000

As we observed, the mean and median values of the columns are almost equal, so I decided to fill in the null values with the median value. The reason I didn't use the mean value is to ensure the meaning of the data and avoid confusion (for example: CouponUsed is 1.75 or OrderCount is 3.008)


In [ ]:
#Choose Numeric Columns
numeric_cols = df_1.select_dtypes(include='number').columns
#Fill with median value
df_1[numeric_cols] = df_1[numeric_cols].fillna(df_1[numeric_cols].median())

In [ ]:
#Check Null Again
df_1.isna().sum()

Churn                          0
Tenure                         0
PreferredLoginDevice           0
CityTier                       0
WarehouseToHome                0
PreferredPaymentMode           0
Gender                         0
HourSpendOnApp                 0
NumberOfDeviceRegistered       0
PreferedOrderCat               0
SatisfactionScore              0
MaritalStatus                  0
NumberOfAddress                0
Complain                       0
OrderAmountHikeFromlastYear    0
CouponUsed                     0
OrderCount                     0
DaySinceLastOrder              0
CashbackAmount                 0
dtype: int64

**Edit the data in column 'PreferredPaymentMode'**

In [ ]:
#Check values in PreferredPaymentMode column
df_1['PreferredPaymentMode'].unique()

array(['Debit Card', 'UPI', 'CC', 'Cash on Delivery', 'E wallet', 'COD',
       'Credit Card'], dtype=object)

This column has values that have the same meaning, so I edited them('Cash on Delivery' = 'COD', 'Credit Card' = 'CC').

In [ ]:
#ReplaceValues
df_1['PreferredPaymentMode'] = df_1['PreferredPaymentMode'].replace({'Cash on Delivery':'COD','Credit Card':'CC'})


In [ ]:
#SaveFile
df_1.to_excel('Eco-churn-clean.xlsx', index = False)